### PR TITLE
[C5][environment] Read configurations and authenticate user in proper environment [publisher][feature][data]

### PR DESCRIPTION
**Pull request by RenukaFernando**
### Proposed changes in this pull request
- Edit `ConfigManager` to read environment configurations from  `org.wso2.carbon.apimgt.rest.api.configurations` REST api
- Authenticate user from environment instance
- Add test-cases for `ConfigManager`
- Make `org.wso2.carbon.apimgt.rest.api.configurations.models.EnvironmentConfigurations` returns one default environment in `List<Environment> environments`
- Implement the feature https://github.com/wso2/carbon-apimgt/issues/4691

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/AuthManager.js [hunk 1]
```diff
@@ -136,18 +133,31 @@ class AuthManager {
...
+    /**
+     * Get login token path from given environment or get default login token path
+     * @param {Object} environment: environment object
+     * @returns {String} loginTokenPath: login token path of the given environment
+     */
+    getTokenEndpoint(environment) {
+        let loginTokenPath;
+        if(environment){
+            loginTokenPath = environment.host + environment.loginTokenPath + this.contextPath;
+        }else{
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/AuthManager.js]: I think there are multiple formatting issues. Please fix them

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/ConfigManager.js [hunk 2]
```diff
@@ -16,22 +16,35 @@
...
 
-    checkRunTimeConfigs() {
+    /**
+     * get configurations from server: deployment.yaml
+     * @returns {Object}: configuration object
+     */
+    static getConfigs() {
+        return {
+            'promised_environments': function () {
+                if(ConfigManager._promised_environments){
```
Comment (ChamNDeSilva) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/ConfigManager.js]: Add a space after if(ConfigManager._promised_environments) before curly bracket [formatting issue]

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/AuthManager.js [hunk 3]
```diff
@@ -136,18 +133,31 @@ class AuthManager {
...
-    getTokenEndpoint() {
-        return this.host + this.token;
+    /**
+     * Get login token path from given environment or get default login token path
+     * @param {Object} environment: environment object
+     * @returns {String} loginTokenPath: login token path of the given environment
+     */
+    getTokenEndpoint(environment) {
+        let loginTokenPath;
+        if(environment){
```
Comment (ChamNDeSilva) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/AuthManager.js]: space required. Formatting issue

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/ConfigManager.js [hunk 4]
```diff
@@ -16,22 +16,35 @@
...
+                    return ConfigManager._promised_environments;
+                }
+                let host = window.location.protocol + "//" + window.location.host;
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/ConfigManager.js]: I think this is equivalent to `location.origin`

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/ConfigManager.js [hunk 5]
```diff
@@ -16,22 +16,35 @@
...
+                let environmentConfigPath = "/configService/environments";
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/ConfigManager.js]: Shall we make this a constant and define as a class variable?

Comment (RenukaFernando) [reply features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/ConfigManager.js]: Fixed with https://github.com/wso2/carbon-apimgt/pull/4692/commits/4510febf6e727f0da1c539e0226e2f1cfb5d283d

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/ConfigManager.js (hunk 4 above)
Comment (RenukaFernando) [reply features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/ConfigManager.js]: Fixed with https://github.com/wso2/carbon-apimgt/pull/4692/commits/4510febf6e727f0da1c539e0226e2f1cfb5d283d

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/AuthManager.js (hunk 3 above)
Comment (RenukaFernando) [reply features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/AuthManager.js]: Fixed with https://github.com/wso2/carbon-apimgt/pull/4692/commits/4510febf6e727f0da1c539e0226e2f1cfb5d283d

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/ConfigManager.js (hunk 2 above)
Comment (RenukaFernando) [reply features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/ConfigManager.js]: Fixed with https://github.com/wso2/carbon-apimgt/pull/4692/commits/4510febf6e727f0da1c539e0226e2f1cfb5d283d

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/AuthManager.js (hunk 1 above)
Comment (RenukaFernando) [reply features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/AuthManager.js]: Fixed with https://github.com/wso2/carbon-apimgt/pull/4692/commits/4510febf6e727f0da1c539e0226e2f1cfb5d283d

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/ConfigManager.js [hunk 6]
```diff
@@ -16,22 +16,48 @@
...
+        return ConfigManager._promised_environments;
     }
-}
\ No newline at end of file
+};
+
+//List of private promised class variables to preserve single instance
+ConfigManager._promised_environments = null;
+
+export default ConfigManager;
```
Comment (TanyaM) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/ConfigManager.js:63]: new line required

**Pull request merged**